cmd/kubeadm/app/cmd/upgrade/plan.go

```diff
@@ -50,25 +55,189 @@ func newCmdPlan(apf *applyPlanFlags) *cobra.Command {
+		TextPrintFlags:     &upgradePlanTextPrintFlags{},
+		TypeSetterPrinter:  printers.NewTypeSetter(outputapischeme.Scheme),
+		OutputFormat:       strings.ToLower(outputFormat),
+	}
+}
+
+// AllowedFormats returns list of allowed output formats
+func (pf *upgradePlanPrintFlags) AllowedFormats() []string {
+	ret := pf.TextPrintFlags.AllowedFormats()
+	return append(ret, pf.JSONYamlPrintFlags.AllowedFormats()...)
+}
+
+// AddFlags receives a *cobra.Command reference and binds
+// flags related to Kubeadm printing to it
+func (pf *upgradePlanPrintFlags) AddFlags(cmd *cobra.Command) {
+	pf.TextPrintFlags.AddFlags(cmd)
+	pf.JSONYamlPrintFlags.AddFlags(cmd)
+	cmd.Flags().StringVarP(&pf.OutputFormat, "output", "o", pf.OutputFormat, fmt.Sprintf("EXPERIMENTAL: Output format. One of: %s.", strings.Join(pf.AllowedFormats(), "|")))
+}
+
+// ToPrinter receives an outputFormat and returns a printer capable of
+// handling format printing.
+// Returns error if the specified outputFormat does not match supported formats.
+func (pf *upgradePlanPrintFlags) ToPrinter() (output.Printer, error) {
+	switch pf.OutputFormat {
+	case output.TextOutput:
+		return pf.TextPrintFlags.ToPrinter(pf.OutputFormat)
+	case output.JSONOutput:
+		return newUpgradePlanJSONYAMLPrinter(pf.TypeSetterPrinter.WrapToPrinter(pf.JSONYamlPrintFlags.ToPrinter(output.JSONOutput)))
+	case output.YAMLOutput:
+		return newUpgradePlanJSONYAMLPrinter(pf.TypeSetterPrinter.WrapToPrinter(pf.JSONYamlPrintFlags.ToPrinter(output.YAMLOutput)))
+	default:
+		return nil, genericclioptions.NoCompatiblePrinterError{OutputFormat: &pf.OutputFormat, AllowedFormats: pf.AllowedFormats()}
+	}
+}
+
+type upgradePlanJSONYamlPrintFlags struct {
+	genericclioptions.JSONYamlPrintFlags
+}
+
+// AllowedFormats returns list of allowed output formats
+func (pf *upgradePlanJSONYamlPrintFlags) AllowedFormats() []string {
+	return []string{output.JSONOutput, output.YAMLOutput}
+}
+
+// upgradePlanJSONYAMLPrinter prints upgrade plan in a JSON or YAML format
+type upgradePlanJSONYAMLPrinter struct {
+	output.ResourcePrinterWrapper
+	Buffer []outputapiv1alpha2.ComponentUpgradePlan
+}
+
+// newUpgradePlanJSONYAMLPrinter creates new upgradePlanJSONYAMLPrinter object
+func newUpgradePlanJSONYAMLPrinter(resourcePrinter printers.ResourcePrinter, err error) (output.Printer, error) {
+	if err != nil {
+		return nil, err
+	}
+	return &upgradePlanJSONYAMLPrinter{ResourcePrinterWrapper: output.ResourcePrinterWrapper{Printer: resourcePrinter}}, nil
+}
+
+// PrintObj is an implementation of ResourcePrinter.PrintObj that adds object to the buffer
+func (p *upgradePlanJSONYAMLPrinter) PrintObj(obj runtime.Object, writer io.Writer) error {
+	item, ok := obj.(*outputapiv1alpha2.ComponentUpgradePlan)
+	if !ok {
+		return fmt.Errorf("expected ComponentUpgradePlan, but got %+v", obj)
+	}
+	p.Buffer = append(p.Buffer, *item)
+	return nil
+}
+
+// Close writes any buffered data and empties list of buffered components
```

Comment:
```suggestion
// Close writes any buffered data and empties the list of buffered components
```

but...printing on close is a bit strange. should we just empty the buffer here?

Reply:
```
// Close empties the list of buffered components
```
An inaccurate naming.